sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubProducerClient.cs

```diff
@@ -401,40 +401,43 @@ public virtual Task<string[]> GetPartitionIdsAsync(CancellationToken cancellatio
             Argument.AssertNotNull(events, nameof(events));
             AssertSinglePartitionReference(options.PartitionId, options.PartitionKey);
 
-            // Determine the transport producer to delegate the send operation to.  Because sending to a specific
-            // partition requires a dedicated client, use (or create) that client if a partition was specified.  Otherwise
-            // the default gateway producer can be used to request automatic routing from the Event Hubs service gateway.
-
-            TransportProducer activeProducer;
-
-            if (string.IsNullOrEmpty(options.PartitionId))
-            {
-                activeProducer = EventHubProducer;
-            }
-            else
-            {
-                // This assertion is intended as an additional check, not as a guarantee.  There still exists a benign
-                // race condition where a transport producer may be created after the client has been closed; in this case
-                // the transport producer will be force-closed with the associated connection or, worst case, will close once
-                // its idle timeout period elapses.
-
-                Argument.AssertNotClosed(IsClosed, nameof(EventHubProducerClient));
-                activeProducer = PartitionProducers.GetOrAdd(options.PartitionId, id => Connection.CreateTransportProducer(id, RetryPolicy));
-            }
+            bool isMessageSent = false;
+            int attempts = 0;
 
             using DiagnosticScope scope = CreateDiagnosticScope();
-
             events = (events as IList<EventData>) ?? events.ToList();
             InstrumentMessages(events);
 
-            try
-            {
-                await activeProducer.SendAsync(events, options, cancellationToken).ConfigureAwait(false);
-            }
-            catch (Exception ex)
+            TransportProducerPool.PooledProducer pooledProducer = PartitionProducerPool.GetPooledProducer(options.PartitionId, PartitionProducerLifespan);
+
+            while (!cancellationToken.IsCancellationRequested
+                   && ++attempts <= MaximumCreateProducerAttempts
+                   && !isMessageSent)
             {
-                scope.Failed(ex);
-                throw;
+                try
+                {
+                    await using var _ = pooledProducer.ConfigureAwait(false);
+
+                    await pooledProducer.TransportProducer.SendAsync(events, options, cancellationToken).ConfigureAwait(false);
+
+                    isMessageSent = true;
```

Comment:
Not sure that we need the `isMessageSent`.... couldn't we just `break` out of the loop on success and save ourselves a stackalloc?

Reply:
I turned the ```break``` into a ```return``` after we included the cancellation logic.